src/main/scala/com/wavesplatform/state2/BlockchainUpdaterImpl.scala

```diff
@@ -178,31 +176,36 @@ class BlockchainUpdaterImpl private(persisted: StateWriter with StateReader,
           log.warn(s"removeAfter nonexistent block $blockId")
           Left(GenericError(s"Failed to rollback to nonexistent block $blockId"))
         case Some(height) =>
-          logHeights(s"Rollback to h=$height started")
           val discardedTransactions = Seq.newBuilder[Transaction]
           discardedTransactions ++= ng.toSeq.flatMap(_.transactions)
           ngState.set(None)
 
-          while (historyWriter.height > height)
-            discardedTransactions ++= historyWriter.discardBlock()
-          if (height < persisted.height) {
-            log.info(s"Rollback to h=$height requested. Persisted height=${persisted.height}, will drop state and reapply blockchain now")
-            persisted.clear()
-            updatePersistedAndInMemory()
-          } else {
-            if (bestLiquidState.height != height) {
-              val persistedPlusBottomHeight = persisted.height + bottomMemoryDiff().heightDiff
-              if (height > persistedPlusBottomHeight) {
-                val newTopDiff = unsafeDiffByRange(composite(persisted, () => bottomMemoryDiff()), persistedPlusBottomHeight + 1, height + 1)
-                topMemoryDiff.set(newTopDiff)
-              } else {
-                topMemoryDiff.set(BlockDiff.empty)
-                if (height < persistedPlusBottomHeight)
-                  bottomMemoryDiff.set(unsafeDiffByRange(persisted, persisted.height + 1, height + 1))
+          if (height < historyWriter.height()) {
```

Comment:
If `historyWriter.heightOf(blockId)` matches `Some(height)`, we actually rollback ng-block if it exists. So technically even if `height == historyWriter.height()` this might be a rollback(at least log-wise).